internal/sandbox/sandbox.go

```diff
@@ -172,40 +165,137 @@ func tokenizePath(path string) []string {
+	components, err := tokenizeMapping(mapping.Mapping)
+	if err != nil {
+		return nil, err
+	}
+	if len(components) == 1 && root != nil {
+		return nil, fmt.Errorf("failed to map root with target %s: root must be mapped first and not more than once", mapping.Target)
+	}
+
+	fileInfo, err := os.Lstat(mapping.Target)
+	if err != nil {
+		return nil, fmt.Errorf("failed to stat %s when mapping %s: %v", mapping.Target, mapping.Mapping, err)
+	}
+
+	if len(components) == 1 {
+		if fileInfo.Mode()&os.ModeType != os.ModeDir {
+			return nil, fmt.Errorf("cannot map file %s at root: must be a directory", mapping.Target)
 		}
+		root = newDir(mapping.Target, fileInfo, mapping.Writable)
+		root.isMapping = true
+	} else {
+		if root == nil {
+			root = newDirEmpty()
+		}
+		dirNode := root.LookupOrCreateDirs(components[1 : len(components)-1])
+		newNode := getOrCreateNode(mapping.Target, fileInfo, mapping.Writable)
+		newNode.SetIsMapping()
+		if err := dirNode.Map(components[len(components)-1], newNode); err != nil {
+			return nil, fmt.Errorf("cannot map %s: %v", mapping.Mapping, err)
+		}
+	}
+	return root, nil
+}
 
-		if len(components) == 1 {
-			if fileInfo.Mode()&os.ModeType != os.ModeDir {
-				return nil, fmt.Errorf("cannot map file %s at root: must be a directory", mapping.Target)
-			}
-			root = newDir(mapping.Target, fileInfo, mapping.Writable)
-			root.isMapping = true
-		} else {
-			if root == nil {
-				root = newDirEmpty()
+// reconfigureMap applies a single Map step to the file system.
+func (f *FS) reconfigureMap(mapping *MappingSpec) error {
+	// TODO(jmmv): We cheat and peek directly into root to get its directory.  The directory
+	// could change with previous implementations of reconfiguration but cannot any longer.
+	// Remove once the Root node indirection is gone.
+	root := f.root.getDir()
+	newRoot, err := applyMapping(root, mapping)
+	if err != nil {
+		return err
+	}
+	if root != newRoot {
+		// applyMapping above has the ability to generate a new root directory when none is
+		// given to it.  We should never encounter that situation here because we do not
+		// allow mapping reconfigurations to modify the root, so ensure this is true.
+		panic("Mapping should not have changed the root but did")
+	}
+	return nil
+}
+
+// reconfigureUnmap applies a single Unmap step to the file system.
+//
+// Note that this only removes the leaf of the given mapping: nested entries created during a map
+// operation are left behind and have to be unmapped explicitly.  E.g. if /foo/bar/baz was mapped
+// on an empty file system, /foo/bar are two intermediate mappings that are not removed when
+// /foo/var/baz is unmapped.  This is for simplicity as we'd otherwise need to track whether a
+// mapping was explicitly created by the user or as a result of another mapping, and also to keep
+// the unmapping algorithm trivial.
+func (f *FS) reconfigureUnmap(server *fs.Server, mapping string) error {
+	components, err := tokenizeMapping(mapping)
```

Comment:
I know that Go encourages short names, but this is really too vague, I think

Can we rename components to pathComponents or to pathTokens? Something that makes it clearer it is a path, anyway.